File: LeetCode_Algorithm/433(Minimum Genetic Mutation)/433.py
'''
LeetCode 433 : Minimum Genetic Mutation

'''
class Solution:
    def minMutation(self, startGene: str, endGene: str, bank: list[str]) -> int:
        # sol1 51ms 5% 13.9MB 77%
        if endGene not in bank:
            return -1
        graph = {
            'A' : ['C', 'G', 'T'],
            'C' : ['A', 'G', 'T'],
            'G' : ['A', 'C', 'T'],
            'T' : ['A', 'C', 'G']
        }
        from collections import deque
        queue = deque([(startGene, 0)])
        seen = set([startGene])
        while queue:
            gene, cnt = queue.popleft()
            if gene == endGene:
                return cnt
            for i in range(8):
                for c in graph[gene[i]]:
                    newGene = gene[:i] + c + gene[i+1:]
                    if newGene in bank and newGene not in seen:
                        seen.add(newGene)
                        queue.append((newGene, cnt+1))
        return -1

File: LeetCode_Algorithm/433(Minimum Genetic Mutation)/433.js
var minMutation = function(startGene, endGene, bank) {
  // sol1 53ms 96% 42.2MB 25%
  if (bank.indexOf(endGene) < 0)
    return -1
  let graph = new Map()
  graph.set('A', ['C', 'G', 'T'])  
  graph.set('C', ['A', 'G', 'T'])
  graph.set('G', ['A', 'C', 'T'])
  graph.set('T', ['A', 'C', 'G'])
  let seen = new Set(); seen.add(startGene)
  let queue = [[startGene, 0]]
  while (queue.length) {
    [gene, cnt] = queue.shift()
    if (gene == endGene)
      return cnt
    for (let i = 0; i < 8; i++) {
      for (let c of graph[gene[i]]) {
        let newGene = gene.slice(0, i) + c + gene.slice(i+1, )
        if (!seen.has(newGene) && bank.indexOf(newGene) >= 0) {
          seen.add(newGene)
          queue.append([newGene, cnt+1])
        }
      }
    }
  }
  return -1
};